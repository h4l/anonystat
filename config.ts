import { HandlerRequest, Matcher, Responder } from "./requests.ts";
import {
  createCollectRequestMatcher,
  DefaultCollectRequestForwardingRule,
  MeasurementIdCollectRequestMetaMatcher,
} from "./rules.ts";

import { z } from "./deps.ts";
import { matchDefaultGA4MPUrls } from "./default.ts";
import { DefaultRequestForwarder, RequestMatchError } from "./proxy.ts";
import { defaultProxyOptions } from "./default.ts";
import { GA4MP_URL } from "./constants.ts";

function normaliseToArray<T>(value: T | T[]): T[] {
  return Array.isArray(value) ? value : [value];
}

function oneOrMore<SchemaT extends z.ZodTypeAny>(schema: SchemaT) {
  return (schema.or(schema.array().min(1))).transform(
    normaliseToArray<z.infer<typeof schema>>,
  );
}

const DataStreamCredentials = z.object({
  measurement_id: z.string(),
  api_secret: z.string(),
});

const DataStreamInOut = z.object({
  in: DataStreamCredentials,
  out: DataStreamCredentials,
});
type DataStreamInOut = z.infer<typeof DataStreamInOut>;

export const DataStreamInOutShorthand = DataStreamCredentials.transform((
  ds,
): DataStreamInOut => ({ in: { ...ds }, out: { ...ds } }));

export const ForwarderConfig = z.object({
  data_stream: oneOrMore(DataStreamInOutShorthand),
  destination: z.string().url().optional().default(GA4MP_URL),
  allow_debug: z.boolean().optional().default(false),
  secret_scrambler_key: z.string().min(8),
});
type ForwarderConfig = z.infer<typeof ForwarderConfig>;

export const Config = z.object({ forward: oneOrMore(ForwarderConfig) });
export type Config = z.infer<typeof Config>;

// TODO: need to validate no duplicate measurement IDs, or handle duplicates

export function createForwarder(
  config: Config,
): Matcher<HandlerRequest, RequestMatchError, Responder> {
  const rules = config.forward.flatMap((forwarder) => {
    const requestForwarder = new DefaultRequestForwarder(defaultProxyOptions);
    return forwarder.data_stream.map((data_stream) => {
      const rule = new DefaultCollectRequestForwardingRule({
        allowedApiSecret: data_stream.in.api_secret,
        destination: {
          measurement_id: data_stream.out.measurement_id,
          api_secret: data_stream.out.api_secret,
          endpoint: forwarder.destination,
        },
        allowDebug: forwarder.allow_debug,
        forwarder: requestForwarder,
      });
      return [data_stream.in.measurement_id, rule] as const;
    });
  });

  const matcher = new MeasurementIdCollectRequestMetaMatcher(rules);

  return createCollectRequestMatcher({
    getRequestMeta: matchDefaultGA4MPUrls,
    collectRequestMetaMatcher: matcher,
  });
}
