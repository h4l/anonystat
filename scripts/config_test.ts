import { ConfigValueEnvarName } from "../config.ts";
import {
  assertEquals,
  assertInstanceOf,
  assertMatch,
  assertSnapshot,
  fromFileUrl,
  Stub,
  stub,
} from "../dev_deps.ts";
import { main } from "./config.ts";

Deno.test("--help shows help", async (t) => {
  using consoleOutput = new ConsoleOutputCapturer();
  await assertExits(main(["--help"]), 0);
  await assertSnapshot(t, consoleOutput.getMergedOutput());
});

Deno.test("loads config from environment", async (t) => {
  using consoleOutput = new ConsoleOutputCapturer();
  using _env = OverrideEnv.withObj({
    ANONYSTAT_DATA_STREAM_IN_MEASUREMENT_ID: "a",
    ANONYSTAT_DATA_STREAM_IN_API_SECRET: "b",
    ANONYSTAT_DATA_STREAM_OUT_MEASUREMENT_ID: "c",
    ANONYSTAT_DATA_STREAM_OUT_API_SECRET: "d",
  });
  await assertExits(main([]), 0);
  await assertSnapshot(t, consoleOutput.getMergedOutput());
});

Deno.test("loads config from argument file", async (t) => {
  using consoleOutput = new ConsoleOutputCapturer();
  using _env = OverrideEnv.withObj({});
  await assertExits(
    main([fromFileUrl(import.meta.resolve("../data/config_minimal.json"))]),
    0,
  );
  await assertSnapshot(t, consoleOutput.getMergedOutput());
});

Deno.test("--format env-vars cannot represent configs with multiple forwarding rules", async (t) => {
  await assertSnapshot(
    t,
    await run({
      format: "env-vars",
      inputFile: "../data/config_multiple.json",
      exitCode: 1,
    }),
  );
});

Deno.test("--format env-vars represents configs with one forwarding rule as individual vars", async (t) => {
  await assertSnapshot(
    t,
    await run({
      format: "env-vars",
      inputFile: "../data/config_single.json",
    }),
  );
});

Deno.test("--format env represents configs with one forwarding rule as individual vars", async (t) => {
  await assertSnapshot(
    t,
    await run({
      format: "env",
      inputFile: "../data/config_single.json",
    }),
  );
});

Deno.test("--format env represents configs with multiple forwarding rule as JSON var", async (t) => {
  await assertSnapshot(
    t,
    await run({
      format: "env",
      inputFile: "../data/config_multiple.json",
    }),
  );
});

Deno.test("--format env-json represents configs with one forwarding rule as JSON var", async (t) => {
  await assertSnapshot(
    t,
    await run({
      format: "env-json",
      inputFile: "../data/config_single.json",
    }),
  );
});

Deno.test("--format json prints indented JSON", async (t) => {
  await assertSnapshot(
    t,
    await run({
      format: "json",
      inputFile: "../data/config_minimal.json",
    }),
  );
});

Deno.test("--format json with --compact prints compact JSON", async (t) => {
  await assertSnapshot(
    t,
    await run({
      format: "json",
      compact: true,
      inputFile: "../data/config_minimal.json",
    }),
  );
});

Deno.test("--format markdown prints a markdown doc with every config format", async (t) => {
  await assertSnapshot(
    t,
    await run({
      format: "markdown",
      compact: true,
      inputFile: "../data/config_minimal.json",
    }),
  );
});

Deno.test("default format is 'env'", async () => {
  const defaultOutput = await run({
    compact: true,
    inputFile: "../data/config_minimal.json",
  });
  const envOutput = await run({
    compact: true,
    format: "env",
    inputFile: "../data/config_minimal.json",
  });
  assertEquals(defaultOutput, envOutput);
});

Deno.test("-e/--override", async (t) => {
  await t.step(
    "non-in/out data stream values override both in and out",
    async () => {
      const a = await run({
        inputFile: "../data/config_single.json",
        override: {
          ANONYSTAT_DATA_STREAM_API_SECRET: "overridden_secret",
          ANONYSTAT_DATA_STREAM_MEASUREMENT_ID: "overridden_id",
        },
      });
      const b = await run({
        inputFile: "../data/config_single.json",
        override: {
          ANONYSTAT_DATA_STREAM_IN_API_SECRET: "overridden_secret",
          ANONYSTAT_DATA_STREAM_OUT_API_SECRET: "overridden_secret",
          ANONYSTAT_DATA_STREAM_IN_MEASUREMENT_ID: "overridden_id",
          ANONYSTAT_DATA_STREAM_OUT_MEASUREMENT_ID: "overridden_id",
        },
      });
      assertEquals(a, b);
    },
  );

  await t.step(
    "overrides replace all values from initial config",
    async () => {
      const vars: Record<ConfigValueEnvarName, string> = {
        ANONYSTAT_ALLOW_DEBUG: "false",
        ANONYSTAT_DATA_STREAM_API_SECRET: "overridden by in/out",
        ANONYSTAT_DATA_STREAM_MEASUREMENT_ID: "overridden by in/out",
        ANONYSTAT_DATA_STREAM_IN_API_SECRET: "s0",
        ANONYSTAT_DATA_STREAM_OUT_API_SECRET: "s1",
        ANONYSTAT_DATA_STREAM_IN_MEASUREMENT_ID: "i0",
        ANONYSTAT_DATA_STREAM_OUT_MEASUREMENT_ID: "i1",
        ANONYSTAT_DESTINATION: "https://override.example.com/mp/collect",
        ANONYSTAT_LISTEN_HOSTNAME: "override.localhost",
        ANONYSTAT_LISTEN_PORT: "9001",
        ANONYSTAT_USER_ID_EXISTING: "keep",
        ANONYSTAT_USER_ID_LIFETIME: "R/2024-01-01/P1W",
        ANONYSTAT_USER_ID_SCRAMBLING_SECRET: "osec",
      };
      const a = await run({
        // config_single has a value for every field
        inputFile: "../data/config_single.json",
        override: vars,
      });
      const b = await run({ env: vars });
      // a overrides all existing values from the file with vars, so it should
      // be equivalent to just loading the vars directly.
      assertEquals(a, b);
    },
  );

  await t.step(
    "overrides are pulled from the environment when no =<value> is used",
    async () => {
      const a = await run({
        inputFile: "../data/config_minimal.json",
        env: {
          ANONYSTAT_LISTEN_HOSTNAME: "foo",
        },
        override: { "*": null },
      });
      const b = await run({
        env: {
          ANONYSTAT_DATA_STREAM_MEASUREMENT_ID: "G-ABCDE12345",
          ANONYSTAT_DATA_STREAM_API_SECRET: "Ab12Ab12Ab12Ab12Ab12Ab",
          ANONYSTAT_LISTEN_HOSTNAME: "foo",
        },
      });
      assertEquals(a, b);
    },
  );
});

type Format = "markdown" | "json" | "env" | "env-json" | "env-vars";

async function run(
  options: {
    format?: Format;
    inputFile?: string;
    exitCode?: number;
    compact?: boolean;
    env?: Partial<Record<ConfigValueEnvarName, string>>;
    override?: Partial<Record<ConfigValueEnvarName | "*", string | null>>;
  },
): Promise<string> {
  const args: string[] = [];
  if (options.format) {
    args.push("--format", options.format);
  }
  if (options.compact) args.push("--compact");
  if (options.inputFile !== undefined) {
    args.push(fromFileUrl(import.meta.resolve(options.inputFile)));
  }
  for (const [name, value] of Object.entries(options.override ?? {})) {
    if (value === undefined) continue;
    if (value === null) args.push("-e", name);
    else args.push("-e", `${name}=${value}`);
  }

  using consoleOutput = new ConsoleOutputCapturer();
  using _env = OverrideEnv.withObj(options.env ?? {});
  await assertExits(
    main(args),
    options.exitCode ?? 0,
  );
  return consoleOutput.getMergedOutput();
}

async function assertExits(
  promise: PromiseLike<void>,
  code: number,
): Promise<void> {
  try {
    return await promise;
  } catch (e) {
    assertInstanceOf(e, Error);
    assertMatch(
      e.message,
      new RegExp(`\\bTest case attempted to exit with exit code: ${code}\\b`),
    );
  }
}

type ConsoleSrc = "log" | "error";
type OutputLine = { src: ConsoleSrc; line: string };

/** Record console.log/error output and prevent writes to stdout/stderr. */
class ConsoleOutputCapturer {
  outputLines: OutputLine[] = [];
  log: Stub<Console, unknown[], void>;
  error: Stub<Console, unknown[], void>;

  constructor() {
    this.log = stub(console, "log", this.getConsoleMethodStub("log"));
    this.error = stub(console, "error", this.getConsoleMethodStub("error"));
  }

  private getConsoleMethodStub(src: ConsoleSrc): (...args: unknown[]) => void {
    return (...args: unknown[]) => {
      this.outputLines.push(
        ...ConsoleOutputCapturer.mergeAndSplitLogArgs(args)
          .map((line) => ({ src, line } as const)),
      );
    };
  }

  private static mergeAndSplitLogArgs(args: unknown[]): string[] {
    return args.map((x) => String(x)).join(" ").split("\n");
  }

  getMergedOutput(): string {
    return this.outputLines.map(({ src, line }) =>
      `${src === "log" ? "out" : "err"}|${line}`
    ).join("\n");
  }

  [Symbol.dispose]() {
    this.log[Symbol.dispose]();
    this.error[Symbol.dispose]();
  }
}

/** Mock Deno.env to return values from a Map. */
class OverrideEnv {
  readonly get: Stub<Deno.Env, [key: string], string | undefined>;
  readonly set: Stub<Deno.Env, [key: string, value: string], void>;
  constructor(readonly env: Map<string, string>) {
    this.get = stub(Deno.env, "get", (key) => this.env.get(key));
    this.set = stub(Deno.env, "set", (key, value) => {
      this.env.set(key, value);
    });
  }

  static with(
    entries?: Iterable<readonly [string, string]> | null,
  ): OverrideEnv {
    return new OverrideEnv(new Map(entries));
  }

  static withObj(object?: Record<string, string>): OverrideEnv {
    return OverrideEnv.with(Object.entries(object ?? {}));
  }

  [Symbol.dispose]() {
    this.get[Symbol.dispose]();
    this.set[Symbol.dispose]();
  }
}
