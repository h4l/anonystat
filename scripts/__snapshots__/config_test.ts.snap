export const snapshot = {};

snapshot[`--help shows help 1`] = `
"out|Validate and print anonystat config.
out|
out|Usage:
out|  Usage: deno run config.ts [-hc] [-f <format>] [<file>] [-e <name>[=<value>|*]]...
out|
out|Configuration is read from environment variables unless <file> is provided, in
out|which case environment variables are ignored.
out|
out|Values in the config loaded from <file> (or the environment) can be overridden
out|by -e <name>[=<value>|*] options naming ANONYSTAT_ environment variables. A glob
out|indicates all matching environment variables override.
out|
out|Examples:
out|
out|  # Load & validate config from env vars, print as NAME=VALUE
out|  config.ts
out|
out|  # Load & validate config from config.json, print as NAME=VALUE
out|  config.ts config.json
out|
out|  # Load from config.json, overriding the scrambling secret from the environment
out|  config.ts config.json -e ANONYSTAT_USER_ID_SCRAMBLING_SECRET
out|
out|  # Load from config.json, overriding the scrambling secret to be 'foo'
out|  config.ts config.json -e ANONYSTAT_USER_ID_SCRAMBLING_SECRET=foo
out|
out|  # Load from config.json, overriding values that are also set as env vars
out|  config.ts config.json -e 'ANONYSTAT_*'
out|
out|  # Load from config.json, output as JSON without indentation
out|  config.ts config.json --format json --compact
out|
out|
out|Arguments:
out|  <file>:
out|    Path of a json[c] config file.
out|
out|Options:
out|  -f <format>, --format <format>:
out|    The representation to print after validating. <format> is:
out|    'env', 'env-json', 'env-vars', 'json', 'markdown'.          [Default: 'env']
out|
out|  -c, --compact:
out|    Don't indent JSON output
out|
out|  -e <name>[=<value>|*], --override <name>[=<value>|*]:
out|    Name an ANONYSTAT_ environment variable that overrides config values
out|    loaded from the environment or <file>. <name> can end with a * to match
out|    multiple, or =<value> to use the provided value instead of the value in the
out|    environment.
out|
out|  -h, --help:
out|    Show this help
out|"
`;

snapshot[`loads config from environment 1`] = `
"out|ANONYSTAT_DATA_STREAM_IN_MEASUREMENT_ID=a
out|ANONYSTAT_DATA_STREAM_IN_API_SECRET=b
out|ANONYSTAT_DATA_STREAM_OUT_MEASUREMENT_ID=c
out|ANONYSTAT_DATA_STREAM_OUT_API_SECRET=d"
`;

snapshot[`loads config from argument file 1`] = `
"out|ANONYSTAT_DATA_STREAM_MEASUREMENT_ID=G-ABCDE12345
out|ANONYSTAT_DATA_STREAM_API_SECRET=Ab12Ab12Ab12Ab12Ab12Ab"
`;

snapshot[`--format env-vars cannot represent configs with multiple forwarding rules 1`] = `"err|Cannot represent config with individual envars: Config contains multiple forward rules"`;

snapshot[`--format env-vars represents configs with one forwarding rule as individual vars 1`] = `
"out|ANONYSTAT_CONFIG_SOURCE=env
out|ANONYSTAT_USER_ID_LIFETIME=R/2024-02-14/P2M
out|ANONYSTAT_USER_ID_EXISTING=replace
out|ANONYSTAT_USER_ID_SCRAMBLING_SECRET=0000000
out|ANONYSTAT_ALLOW_DEBUG=true
out|ANONYSTAT_DESTINATION=https://example.com/mp/collect
out|ANONYSTAT_LISTEN_HOSTNAME=1.2.3.4
out|ANONYSTAT_LISTEN_PORT=9000
out|ANONYSTAT_DATA_STREAM_IN_MEASUREMENT_ID=foo
out|ANONYSTAT_DATA_STREAM_IN_API_SECRET=secret123
out|ANONYSTAT_DATA_STREAM_OUT_MEASUREMENT_ID=G-ABCDE12345
out|ANONYSTAT_DATA_STREAM_OUT_API_SECRET=Ab12Ab12Ab12Ab12Ab12Ab"
`;

snapshot[`--format env represents configs with one forwarding rule as individual vars 1`] = `
"out|ANONYSTAT_USER_ID_LIFETIME=R/2024-02-14/P2M
out|ANONYSTAT_USER_ID_EXISTING=replace
out|ANONYSTAT_USER_ID_SCRAMBLING_SECRET=0000000
out|ANONYSTAT_ALLOW_DEBUG=true
out|ANONYSTAT_DESTINATION=https://example.com/mp/collect
out|ANONYSTAT_LISTEN_HOSTNAME=1.2.3.4
out|ANONYSTAT_LISTEN_PORT=9000
out|ANONYSTAT_DATA_STREAM_IN_MEASUREMENT_ID=foo
out|ANONYSTAT_DATA_STREAM_IN_API_SECRET=secret123
out|ANONYSTAT_DATA_STREAM_OUT_MEASUREMENT_ID=G-ABCDE12345
out|ANONYSTAT_DATA_STREAM_OUT_API_SECRET=Ab12Ab12Ab12Ab12Ab12Ab"
`;

snapshot[`--format env represents configs with multiple forwarding rule as JSON var 1`] = `
'out|ANONYSTAT_CONFIG_SOURCE=json
out|ANONYSTAT_CONFIG={"forward":[{"data_stream":[{"in":{"measurement_id":"mid1","api_secret":"defaultsecret999"},"out":{"measurement_id":"midout1","api_secret":"defaultsecret888"}},{"in":{"measurement_id":"mid2","api_secret":"defaultsecret999"},"out":{"measurement_id":"midout2","api_secret":"defaultsecret888"}}],"user_id":{"lifetime":"2 weeks","scrambling_secret":"foobarbaz4"},"allow_debug":true,"destination":"https://lol/foo/123"},{"data_stream":{"in":{"measurement_id":"mid3","api_secret":"defaultsecret999"},"out":{"measurement_id":"midout1","api_secret":"defaultsecret888"}},"destination":"https://example.com/foo/123"}],"listen":{"hostname":"1.2.3.4","port":9000}}'
`;

snapshot[`--format env-json represents configs with one forwarding rule as JSON var 1`] = `
'out|ANONYSTAT_CONFIG_SOURCE=json
out|ANONYSTAT_CONFIG={"forward":{"data_stream":{"in":{"measurement_id":"foo","api_secret":"secret123"},"out":{"measurement_id":"G-ABCDE12345","api_secret":"Ab12Ab12Ab12Ab12Ab12Ab"}},"user_id":{"existing":"replace","lifetime":{"count":2,"unit":"months","from":"2024-02-14"},"scrambling_secret":"0000000"},"allow_debug":true,"destination":"https://example.com/mp/collect"},"listen":{"hostname":"1.2.3.4","port":9000}}'
`;

snapshot[`--format json prints indented JSON 1`] = `
'out|{
out|  "forward": {
out|    "data_stream": {
out|      "api_secret": "Ab12Ab12Ab12Ab12Ab12Ab",
out|      "measurement_id": "G-ABCDE12345"
out|    }
out|  }
out|}'
`;

snapshot[`--format json with --compact prints compact JSON 1`] = `'out|{"forward":{"data_stream":{"api_secret":"Ab12Ab12Ab12Ab12Ab12Ab","measurement_id":"G-ABCDE12345"}}}'`;

snapshot[`--format markdown prints a markdown doc with every config format 1`] = `
'out|# Anonystat Config Information
out|
out|This is your currently-active configuration, validated and presented in
out|different ways.
out|
out|## Loaded config
out|
out|The config loaded from provided environment variables, normalised with defaults
out|applied.
out|
out|\`\`\`json
out|{
out|  "forward": [
out|    {
out|      "data_stream": [
out|        {
out|          "in": {
out|            "measurement_id": "G-ABCDE12345",
out|            "api_secret": "Ab12Ab12Ab12Ab12Ab12Ab"
out|          },
out|          "out": {
out|            "measurement_id": "G-ABCDE12345",
out|            "api_secret": "Ab12Ab12Ab12Ab12Ab12Ab"
out|          }
out|        }
out|      ],
out|      "destination": "https://www.google-analytics.com/mp/collect",
out|      "allow_debug": false,
out|      "user_id": {
out|        "scrambling_secret": null,
out|        "lifetime": {
out|          "count": 1,
out|          "unit": "months"
out|        },
out|        "existing": "scramble"
out|      }
out|    }
out|  ],
out|  "listen": {
out|    "port": 8000,
out|    "hostname": "127.0.0.1"
out|  }
out|}
out|\`\`\`
out|
out|## Individual environment variables (\`--format env\`)
out|
out|\`\`\`console
out|ANONYSTAT_CONFIG_SOURCE=env
out|ANONYSTAT_DATA_STREAM_MEASUREMENT_ID=G-ABCDE12345
out|ANONYSTAT_DATA_STREAM_API_SECRET=Ab12Ab12Ab12Ab12Ab12Ab
out|\`\`\`
out|
out|
out|## Single environment variable containing JSON config (\`--format env-json\`)
out|
out|\`\`\`console
out|ANONYSTAT_CONFIG_SOURCE=json
out|ANONYSTAT_CONFIG={"forward":{"data_stream":{"api_secret":"Ab12Ab12Ab12Ab12Ab12Ab","measurement_id":"G-ABCDE12345"}}}
out|\`\`\`
out|
out|## JSON config (\`--format json\`)
out|
out|The loaded config with defaults removed and unnecessary arrays removed.
out|
out|\`\`\`json
out|{
out|  "forward": {
out|    "data_stream": {
out|      "api_secret": "Ab12Ab12Ab12Ab12Ab12Ab",
out|      "measurement_id": "G-ABCDE12345"
out|    }
out|  }
out|}
out|\`\`\`
out|
out|To load configuration from a file containing the above JSON content, set these
out|environment variables:
out|
out|\`\`\`console
out|ANONYSTAT_CONFIG_SOURCE=file
out|ANONYSTAT_CONFIG_FILE=/example/path/config.json
out|\`\`\`
out|
out|'
`;
